ui/src/form_helpers.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  clearResponse,
+  createForm,
+  FieldArrayPath,
+  FieldArrayProps,
+  FieldPathValue,
+  FieldProps,
+  FieldValues,
+  FormOptions,
+  FormProps,
+  FormStore,
+  FieldPath,
+  MaybeValue,
+  PartialKey,
+  reset,
+  ResponseData,
+  setResponse,
+  getValues,
+  PartialValues,
+} from '@modular-forms/solid'
+import {
+  Accessor,
+  createMemo,
+  createSignal,
+  JSXElement,
+  Setter,
+} from 'solid-js'
+import { getErrorMessage } from './api'
+
+/**
+ * The form state can be used to keep track of a piece of data for a form.
+ * It provides the sate of the form read and write access to the form data
+ * bypassing user input as wel as an onSubmit callback that should be called
+ * when the form is submitted.
+ *
+ */
+interface FormState<TType extends FieldValues, TResponse extends ResponseData> {
+  /**
+   * The form state gives access to information about the state of the forms data.
+   */
+  state: FormStore<TType, TResponse>
+  /**
+   * The callback that will be called when the form is submitted by the user.
+   */
+  onSubmit: (values: TType) => void
+  /*
+   * A setter to for write access to the data bypassing user input.
+   */
+  setter: Setter<Partial<TType | undefined>>
+  accessor: Accessor<PartialValues<TType>>
+  /*
+   * The form components
+   */
+  components: {
+    Form: (props: Omit<FormProps<TType, TResponse>, 'of'>) => JSXElement
+    Field: <TFieldName extends FieldPath<TType>>(
+      props: FieldPathValue<TType, TFieldName> extends MaybeValue<string>
+        ? PartialKey<
+            Omit<FieldProps<TType, TResponse, TFieldName>, 'of'>,
+            'type'
+          >
+        : Omit<FieldProps<TType, TResponse, TFieldName>, 'of'>
+    ) => JSXElement
+    FieldArray: <TFieldArrayName extends FieldArrayPath<TType>>(
+      props: Omit<FieldArrayProps<TType, TResponse, TFieldArrayName>, 'of'>
+    ) => JSXElement
+  }
+}
+
+/**
+ *  Creates a from state
+ *
+ */
```

Comment:
I really like it, but it is also a bit of magic. Maybe an example doc would be nice